File: Scala/src/test/scala/org/revian/MyDequeTest.scala
package org.revian

import org.scalatest.funsuite.AnyFunSuite

class MyDequeTest extends AnyFunSuite {

  val myDeque = new MyDeque

  test("Add and remove from front of deque") {
    myDeque.addFirst(10)
    assert(myDeque.size() == 1)
    assert(myDeque.removeFirst() == 10)
  }

  test("Add and remove from back of deque") {
    myDeque.addLast(20)
    assert(myDeque.size() == 1)
    assert(myDeque.removeLast() == 20)
  }

  test("Remove from empty deque should throw exception") {
    assertThrows[NoSuchElementException] {
      myDeque.removeFirst()
    }
  }
}

File: Java/src/test/java/org/revian/MyDequeTest.java
package org.revian;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class MyDequeTest {
    private MyDeque deque;

    @BeforeEach
    void setUp() {
        deque = new MyDeque();
    }

    @Test
    void testAddFirstAndRemoveFirst() {
        deque.addFirst(1);
        deque.addFirst(2);
        assertEquals(2, deque.removeFirst());
        assertEquals(1, deque.removeFirst());
    }

    @Test
    void testAddLastAndRemoveLast() {
        deque.addLast(1);
        deque.addLast(2);
        assertEquals(2, deque.removeLast());
        assertEquals(1, deque.removeLast());
    }

    @Test
    void testSize() {
        deque.addFirst(1);
        deque.addLast(2);
        assertEquals(2, deque.size());
    }

    @Test
    void testRemoveFirstEmptyDeque() {
        assertThrows(IllegalStateException.class, () -> deque.removeFirst());
    }

    @Test
    void testRemoveLastEmptyDeque() {
        assertThrows(IllegalStateException.class, () -> deque.removeLast());
    }
}

